### PR TITLE
Implement FILE_SIZE_BYTES

### PR DESCRIPTION
**Pull request by lnkuiper**
This PR implements the `FILE_SIZE_BYTES` parameter for `COPY ... TO` functions for all supported formats (JSON, Parquet, CSV).

If this parameter is set, we create a directory instead of a single file, just like our `PER_THREAD_OUTPUT` parameter. If the written file exceeds the set limit, which can be set in bytes, e.g., `1000`, or in human-readable format, e.g., `'1kb'`, we create a new file in the directory and continue appending to that one. This parameter works in combination with `PER_THREAD_OUTPUT`.

Our writers write in chunks at a time, or for Parquet, even entire row groups at a time. Therefore, the file size will exceed the set parameter in many cases. However, with a sufficient size, the file size should be pretty accurate. When multiple threads are writing to the same file, we may overshoot a bit more.

### CODE REVIEW

**Comment by hannes** [comment]
We should document the fact that the limit is not hard limit but more an approximation once this is merged @szarnyasg

File: src/execution/operator/persistent/physical_copy_to_file.cpp [hunk 1]
```diff
@@ -72,8 +152,72 @@ SinkResultType PhysicalCopyToFile::Sink(ExecutionContext &context, DataChunk &ch
...
+		return SinkResultType::NEED_MORE_INPUT;
+	}
+
+	// FILE_SIZE_BYTES is set, but threads write to the same file, need to synchronize finalizing somehow
+	auto &gstate = g.global_state;
+
+	bool take_ownership = false;
+	unique_ptr<GlobalFunctionData> owned_gstate;
+	while (true) {
+		if (owned_gstate) {
```
Comment (Mytherin) [inline src/execution/operator/persistent/physical_copy_to_file.cpp:178]: Maybe we can clean this up using a shared/exclusive lock (see `StorageLock` in `src/include/duckdb/storage/storage_lock.hpp`) - the shared lock can be used when calling `copy_to_sink`, the exclusive lock when calling `copy_to_finalize`.

File: src/execution/operator/persistent/physical_copy_to_file.cpp [hunk 2]
```diff
@@ -150,16 +294,17 @@ SinkCombineResultType PhysicalCopyToFile::Combine(ExecutionContext &context, Ope
...
-		                         *l.local_state);
-
+	} else if (function.copy_to_combine) {
 		if (per_thread_output) {
+			// For PER_THREAD_OUTPUT, we can combine/finalize immediately
+			function.copy_to_combine(context, *bind_data, *l.global_state, *l.local_state);
 			function.copy_to_finalize(context.client, *bind_data, *l.global_state);
+		} else if (file_size_bytes.IsValid()) {
+			// File in global state may change with FILE_SIZE_BYTES, need to grab lock
+			lock_guard<mutex> guard(g.lock);
```
Comment (Mytherin) [inline src/execution/operator/persistent/physical_copy_to_file.cpp:304]: Shared lock?

File: src/execution/operator/persistent/physical_copy_to_file.cpp [hunk 3]
```diff
@@ -177,78 +322,15 @@ SinkFinalizeType PhysicalCopyToFile::Finalize(Pipeline &pipeline, Event &event,
 		function.copy_to_finalize(context, *bind_data, *gstate.global_state);
 
 		if (use_tmp_file) {
-			D_ASSERT(!per_thread_output); // FIXME
-			D_ASSERT(!partition_output);  // FIXME
+			D_ASSERT(!per_thread_output);         // FIXME
+			D_ASSERT(!partition_output);          // FIXME
+			D_ASSERT(!file_size_bytes.IsValid()); // FIXME
```
Comment (Mytherin) [inline src/execution/operator/persistent/physical_copy_to_file.cpp:327]: Can you remove the FIXME's here I don't think they should be fixed

File: src/execution/operator/persistent/physical_copy_to_file.cpp [hunk 4]
```diff
@@ -40,10 +47,83 @@ class CopyToFunctionLocalState : public LocalSinkState {
...
 
+unique_ptr<GlobalFunctionData> PhysicalCopyToFile::CreateFileState(ClientContext &context, GlobalSinkState &sink,
+                                                                   bool already_locked) const {
+	auto &g = sink.Cast<CopyToFunctionGlobalState>();
+	idx_t this_file_offset;
+	if (already_locked) {
+		this_file_offset = g.last_file_offset++;
+	} else {
+		lock_guard<mutex> glock(g.lock);
+		this_file_offset = g.last_file_offset++;
```
Comment (Mytherin) [inline src/execution/operator/persistent/physical_copy_to_file.cpp:58]: Maybe we can just turn this into an atomic?

File: test/sql/copy/file_size_bytes.test [hunk 5]
```diff
@@ -0,0 +1,159 @@
+# name: test/sql/copy/file_size_bytes.test
```
Comment (Mytherin) [inline test/sql/copy/file_size_bytes.test:1]: Maybe add a larger `.test_slow` to test this behavior that creates many files?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good - some comments below:

**Comment by lnkuiper** [comment]
I've implemented the feedback, looks much cleaner with a shared lock. Thanks!

**Comment by lnkuiper** [comment]
I think this is ready to go, ODBC CI failure is a failed download

**Comment by Mytherin** [comment]
Thanks!

**Comment by szarnyasg** [comment]
Documented via https://github.com/duckdb/duckdb-web/pull/1819

**Pull request merged**